internal/charmstore/resources_test.go

```diff
@@ -0,0 +1,256 @@
+		"~bob/wordpress|res|1",
+		"~bob/wordpress|res2|0",
+		"~bob/wordpress|res2|1",
+		"~bob/mysql|res|0",
+		"~alice/wordpress|res|0",
+	} {
+		parts := strings.SplitN(r, "|", 3)
+		rev, err := strconv.Atoi(parts[2])
+		c.Assert(err, jc.ErrorIsNil)
+		err = store.DB.Resources().Insert(&mongodoc.Resource{
+			BaseURL:  charm.MustParseURL(parts[0]),
+			Name:     parts[1],
+			Revision: rev,
+		})
+		c.Assert(err, jc.ErrorIsNil)
+	}
+	for i, test := range newResourceQueryTests {
+		c.Logf("%d. %s", i, test.about)
+		q := newResourceQuery(test.url, test.name, test.revision)
+		var results []*mongodoc.Resource
+		err := store.DB.Resources().Find(q).All(&results)
+		c.Assert(err, jc.ErrorIsNil)
+		sortResources(test.expectResources)
+		sortResources(results)
+		c.Assert(results, jc.DeepEquals, test.expectResources)
+	}
+}
+
+func (s *resourceSuite) TestListResourcesCharmWithResources(c *gc.C) {
+	store := s.newStore(c, false)
+	defer store.Close()
+	channel := params.StableChannel
+	curl := charm.MustParseURL("cs:~charmers/xenial/starsay-3")
+	entity, ch := addCharm(c, store, curl)
+	expected := addResources(c, store, curl, channel, entity, ch)
+	sortResources(expected)
+
+	docs, err := store.ListResources(entity, channel)
+	c.Assert(err, jc.ErrorIsNil)
+
+	checkResourceDocs(c, docs, expected)
+}
+
+func (s *resourceSuite) TestListResourcesCharmWithoutResources(c *gc.C) {
+	store := s.newStore(c, false)
+	defer store.Close()
+	channel := params.StableChannel
+	curl := charm.MustParseURL("cs:~charmers/precise/wordpress-23")
+	entity, _ := addCharm(c, store, curl)
+
+	resources, err := store.ListResources(entity, channel)
+	c.Assert(err, jc.ErrorIsNil)
+
+	c.Check(resources, gc.HasLen, 0)
+}
+
+func (s *resourceSuite) TestListResourcesBundle(c *gc.C) {
+	store := s.newStore(c, false)
+	defer store.Close()
+	channel := params.StableChannel
+	curl := charm.MustParseURL("cs:~charmers/bundle/wordpress-simple-0")
+	entity := addBundle(c, store, curl)
+
+	_, err := store.ListResources(entity, channel)
+
+	c.Check(err, gc.ErrorMatches, `bundles do not have resources`)
+}
+
+func (s *resourceSuite) TestListResourcesResourceNotFound(c *gc.C) {
+	store := s.newStore(c, false)
+	defer store.Close()
+	channel := params.StableChannel
+	curl := charm.MustParseURL("cs:~charmers/xenial/starsay-3")
+	entity, ch := addCharm(c, store, curl)
+	expected := extractResources(c, curl, ch)
+	sortResources(expected)
+	expected[0] = &mongodoc.Resource{
+		BaseURL: expected[0].BaseURL,
+		Name:    expected[0].Name,
+	}
+	expected[1].Revision = addResource(c, store, entity, channel, expected[1], nil)
+	expected[2].Revision = addResource(c, store, entity, channel, expected[2], nil)
+
+	docs, err := store.ListResources(entity, channel)
+	c.Assert(err, jc.ErrorIsNil)
+
+	checkResourceDocs(c, docs, expected)
+}
+
+func addResources(c *gc.C, store *Store, curl *charm.URL, channel params.Channel, entity *mongodoc.Entity, ch *charm.CharmDir) []*mongodoc.Resource {
```

Comment:
Doc comment please. And like I said below, I'd prefer if this didn't rely on files in the testing charm repo.